### DISCUSSION TITLE
Thin-client (DELL Wyse) - Audio not working

### DISCUSSION THREAD

**Discussion by ldup** [neutrinolabs/xrdp#2600] [Q&A]
Hi, 
Using Ubuntu 22.04.2 LTS and installation script 1.4.6 (http://c-nergy.be/blog/?p=18725).

The installation is OK (except that xrdp 0.9.17 is fetched by the script instead of the latest 0.9.21).

Audio is working well from a Windows client but from a Dell-Wyse Thin-client (ThinOS 8.6) it is not. 
The device says "cannot open device file".
Those Thin-Clients are very common and work very well as RDP client of all flavors of Windows since ages including audio and video redirection.
Any hints ? Is it possible to change the output audio codec ?

**Comment by matt335672** [comment]
The codec used is a result of a negotiation between the server and the client, described [here](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpea/8feb4dc0-42eb-428f-98ad-f579daad701c)

You should find a list of the formats received from the client in the chansrv log file in `~/.local/share/xrdp/xrdp-chansrv.<displaynum>.log`. Looking at the latest source, I think we'll only support PCM audio, 2 channels, 44100 samples/sec, 16 bits/sample. If you can configure the client to do that you're in with a chance.

**Comment by ldup** [comment]
Unfortunately the log provides no clue. Maybe I have to enable a verbose mode or something ?
The client side (ThinOS 8) is closed source and can't be changed. As I said ThinOS has been widely used for ages (since RDP 6 at least.) w/o such issue.

Note that there's a new release (ThinOS 9) completely rewritten and now based on Linux + Microsoft AVD client. 
The thing is that NLA authentication is enforced. So I can't test it on xrdp hosts. I did not see an option to disable NLA. I asked the Dell-Wyse support.

**Comment by matt335672** [comment]
Check you're running the chansrv at log level info or higher. See `ChansrvLogging]` in `/etc/xrdp/sesman.ini`. Also check the date on chansrv.log to make sure it's current.

When the client connects, you should see some info messages along these lines:-

```
 sound_process_output_format:
       wFormatNo       0
       wFormatTag      WAVE_FORMAT_PCM
       nChannels       2
       nSamplesPerSec  44100
       nAvgBytesPerSec 176400
       nBlockAlign     4
       wBitsPerSample  16
       cbSize          0
```

If you don't see those, the possibilities I can see are:-
1) the client doesn't support PCM audio (unlikely)
2) the client isn't advertising the "rdpsnd" channel (possible)
3) you haven't got `rdpsnd=true` in the `[Channels]` section of `/etc/xrdp/xrdp.ini`.
4) you haven't got `allow_channels=true` in the `[Globals]` section of `/etc/xrdp/xrdp.ini`.

**Reply by ldup** [reply]
Sorry I don't get such logging information, even set at the DEBUG level (and restart of the entire VM)

There's only a few lines :

>20230320-16:34:33] [DEBUG] sound_process_wave_confirm: wTimeStamp 27732, cConfirmedBlockNo 67 time diff 1629

As I said, from Windows Client, sound is OK, so other parameters are OK.
It is only something that the Wyse Thinclient does not understand  (and that never occurred so far with Windows host)

Generally speaking there is a trend, I think, in moving Desktops computer into virtualized computers, either cloud on on-premise. Just like servers have been moved.
This requires a good "Remote Desktop" support in terms of multimedia : Sound & video. especially when Thin-clients are involved.
There are various actors accompanying that trend/transition : Microsoft (AVD/RDP), Citrix (Workspace), VmWare (Blast), Teradici (PCOIP). 
Zoom provides a VDI flavor w/ a companion local plugin. Available for Linux, ThinOS
Thanks for the work on Ubtuntu and XRDP and be prepared to have more requests such as video redirection & UDP.

**Comment by matt335672** [comment]
We've had interesting Wyse issues before, where the client hasn't conformed to the specification.

Your logging message suggests to me that the client is skipping the initialisation stage in the spec and just sending sound data.

To really work out what is going on, we'll need a debug build of xrdp, or a suitable thin client. Are you in a position to build xrdp from source maybe?

**Reply by ldup** [reply]
There's a new ThinOS 9 release expected this month. The ability to disable LNA should be there. 
I will be able to test it w/ Ubuntu.
The AVD/RDP client part in there is Microsoft's AVD (for linux). Hopefully Microsoft conforms to its own specification ;-)
I'll keep you posted.

**Reply by matt335672** [reply]
That sounds interesting - please do update this issue.

I just had a look to see if a ThinOS VM was available, but it seems not.

**Reply by ldup** [reply]
Hi,

Using ThinOS 9.4 (2303) audio is working.

Audio alone is OK but the "multimedia" experience when connected to Ubuntu is really bad though.
Watching youtube even in "degraded" condition is impossible.

I would have said that is has to do with UDP, but it turns out that ThinOS 9 + Micosoft AVD does not leverage UDP...(Which is flabbergasting from Microsoft BTW).

OTOH, using ThinOS 9 as client to a Windows VM, the multimedia experience is fair, even w/o UDP .

So the culprit is elsewhere.

Maybe it has to do with video and audio codecs that are too CPU intensive on the XRDP side. Or the audio/video sync (if any) is too aggressive and degrades throughput.

We are on a LAN, so network bandwidth should not be a problem.

So, at the moment a move from a Windows VM to a Ubuntu VM as primary environment for some users is not an option.
Users will still have to log first on a Windows VM (also for camera support through RDP).

Regards

(BTW, I don't know what happened to DELL-Wyse staff who developed the proprietary RDP implementation, now that they are using Microsoft AVD.
Maybe there are a few people out there to "recruit" for XRP..)

**Comment by matt335672** [comment]
If you can configure your client to use a connection type of "LAN" rather than "auto detect" you may be able to leverage RFX which should improve things somewhat. That's the best you can do now, but it's worth investigating. 

RDP over UDP requires a lot of rework of existing transports to be implemented. The FreeRDP project is looking into it [at the moment](/FreeRDP/FreeRDP/issues/4978), so it will be interesting to see what they find. *At the moment there's no evidence at all that any investment in this area will be met with corresponding performance gains* - there's interest in the feature but it doesn't seem to be backed up with any meaningful analysis that I can find.
